src/style-spec/expression/compound_expression.js

```diff
@@ -2,37 +2,37 @@
 
 const { toString } = require('./types');
 const ParsingContext = require('./parsing_context');
+const EvaluationContext = require('./evaluation_context');
 const assert = require('assert');
 
-import type { Expression }  from './expression';
-import type CompilationContext  from './compilation_context';
+import type { Expression } from './expression';
 import type { Type } from './types';
 
 type Varargs = {| type: Type |};
 type Signature = Array<Type> | Varargs;
-type Compile = (ctx: CompilationContext, args: Array<Expression>) => string;
-type Definition = [Type, Signature, Compile] |
-    {|type: Type, overloads: Array<[Signature, Compile]>|};
+type Evaluate = Function;
+type Definition = [Type, Signature, Evaluate] |
+    {|type: Type, overloads: Array<[Signature, Evaluate]>|};
 
 class CompoundExpression implements Expression {
     key: string;
     name: string;
     type: Type;
-    compileFromArgs: Compile;
+    _evaluate: Evaluate;
     args: Array<Expression>;
 
     static definitions: { [string]: Definition };
 
-    constructor(key: string, name: string, type: Type, compileFromArgs: Compile, args: Array<Expression>) {
+    constructor(key: string, name: string, type: Type, evaluate: Evaluate, args: Array<Expression>) {
         this.key = key;
         this.name = name;
         this.type = type;
-        this.compileFromArgs = compileFromArgs;
+        this._evaluate = evaluate;
         this.args = args;
     }
 
-    compile(ctx: CompilationContext) {
-        return this.compileFromArgs(ctx, this.args);
+    evaluate(ctx: EvaluationContext) {
+        return this._evaluate.apply(ctx, this.args.map(arg => arg.evaluate(ctx)));
```

Comment:
Evaluating arguments ahead of time means that `&&` and `||` won't stop early, i.e. `["||", true, ["error", ""]]` would produce an error rather than evaluating to `true`.  We should make these a special form (and I should have had tests for this).  I'm gonna rebase the `expressions-filters` to pick up these changes, and can pick this up as part of that work.

Reply:
👍  Filed https://github.com/mapbox/mapbox-gl-js/issues/5403 to track this.